[New feature] External subtitles over UPnP

7pepo7 opened the pull request:
Like in title, it allows to play video with external subtitle over UPnP.

Supported subtitle file extensions: txt, srt, ssa, ass, sub, smi. (most
popular ones)

Subtitle files are searched like in internal player (in video file
folder and in folders specified in options). If more than one subtitle
file are found (multiple language versions), Kodi takes the one with
language choosed in options (prefered language for subtitles), otherwise
first found.

Subtitles are added as 2 resources, 2 sec resources and 3 attributes added to
video file resource, for greater compatibility with more UPnP devices.
Device take the last one it could handle, and skip ones it can't
"understand":

&lt;res protocolInfo="http-get:_:video/mp4:_" xmlns:pv="http://www.pv.com/pvns/" pv:subtitleFileUri="http://192.168.1.13:1722/video.srt" pv:subtitleFileType="SRT"&gt;http://192.168.1.13:1722/video.mp4 &lt;/res&gt;
&lt;res protocolInfo="http-get:_:text/srt:_"&gt; http://192.168.1.13:1722/video.srt &lt;/res&gt;
&lt;res protocolInfo="http-get:_:smi/caption:_"&gt; http://192.168.1.13:1722/video.srt &lt;/res&gt;
&lt;sec:CaptionInfoEx sec:type="srt"&gt;http://192.168.1.13:1722/video.srt&lt;/sec:CaptionInfoEx&gt;
&lt;sec:CaptionInfo sec:type="srt"&gt;http://192.168.1.13:1722/video.srt&lt;/sec:CaptionInfo&gt;

Samsung devices get subtitles uri from "CaptionInfo.sec" field in http response header for movie file request (it contains "getCaptionInfo.sec" field).

This all should be (and probably it would be) changed, when Profile Manager (for DLNA devices) will be available in Kodi.

Montellese commented on lib/libUPnP/Platinum/Source/Devices/MediaServer/PltMediaItem.cpp:
```diff
@@ -528,7 +528,21 @@ PLT_MediaObject::ToDidl(NPT_UInt64 mask, NPT_String& didl)
+            while (entry)
+            {
+                didl += " ";
+                PLT_Didl::AppendXmlEscape(didl, (NPT_String&)(*entry)->GetKey());
```
Is the explicit cast to `NPT_String&` necessary? `m_CustomData` already is a `NPT_Map<NPT_String, NPT_String>` so `GetKey()` and `GetValue()` should already return `NPT_String`.

Montellese commented on xbmc/network/upnp/UPnPInternal.cpp:
```diff
@@ -606,6 +610,82 @@ BuildObject(CFileItem&                    item,
+        std::vector<std::string> subtitles;
+        CUtil::ScanForExternalSubtitles(file_path.GetChars(), filenames);
+
+        NPT_String ext;
```
Please use `std::string` and `StringUtils`. `NPT_String` should only be used where it comes out of the Platinum SDK.

Montellese commented on xbmc/network/upnp/UPnPInternal.cpp:
```diff
@@ -606,6 +610,82 @@ BuildObject(CFileItem&                    item,
+        else if (!subtitles.empty())
+        {
+            /* trying to find subtitle with prefered language settings */
+            std::string preferedLanguage = (CSettings::Get().GetSetting("locale.subtitlelanguage"))->ToString();
```
This logic probably already exists somewhere (probably in DVDPlayer) so would be nice if it could be refactored so it doesn't have to be duplicated.

Montellese commented on xbmc/network/upnp/UPnPInternal.cpp:
```diff
@@ -606,6 +610,82 @@ BuildObject(CFileItem&                    item,
+                }
+            }
+			/* if not found subtitle with prefered language, get the first one */
+            if (subtitlePath == "")
```
`subtitlePath.empty()`

Montellese commented on xbmc/network/upnp/UPnPInternal.cpp:
```diff
@@ -606,6 +610,82 @@ BuildObject(CFileItem&                    item,
+            }
+        }
+
+        if (subtitlePath != "")
```
`!subtitlePath.empty()`

Montellese commented:
I don't use subtitles but this looks like it will make a lot of people happy.

Please split this into three commits:
1. Only the changes to `lib/libUPnP/Platinum/`
2. Add a patch file of the previous commit to `lib/libUPnP/patches`
3. All the other changes to our UPnP integration

7pepo7 commented:
Thanks for review. I'll try to do changes as quickly as possible. But I have few questions. I'm new in github. I work in Mercurial with SourceTree and mostly for sync my code between work and home, not for team work, so I didn't make pull requests and patches. Could you explain me how to split this to three commits: make new 3 commits and new pull request or something else? What with this pull request? And how to make a patch and add it to remote folder. Sorry for neewbe questions, but I prefer to ask than to do something for which you want to kill me :D Please for understanding :D

Montellese commented:
No need for hurry. I can only provide you the git bash commands as I'm not using any GUI for git. You need to be on the branch for this PR i.e. `master-subtitles`.
1. `git reset --soft HEAD^` will remove the commit but keep the changes staged
2. `git reset HEAD` will unstage the changed files but still keep the changes
3. Make the necessary changes to the source code and make sure it still compiles and works
4. Use `git add` for every changed file in `lib/libUPnP/Platinum/`
5. Use `git commit -m "<message>"` to commit those changes
6. `git format-patch -1` will create a `.patch` file in your working directory of the previous commit.
7. Move that file to `lib/libUPnP/patches` and change the leading number accordingly to be the last in the list of the patches
8. Use `git add` on the patch file
9. Use `git commit -m "<message>"` to commit the patch file
10. Use `git add` to add the remaining files/changes
12. Use `git commit -m "<message>"` to commit those changes
13. `git push -f origin master-subtitles` will force push your changes to your remote branch and this PR will automatically be updated.

7pepo7 commented:
Wow, thank you for so detailed explanation. I'll make changes as soon as I find some time. Thanks again.

7pepo7 committed:
platinum Changes to external subtitles over UPnP works

Added custom data to resources field - needed to add some attributes to res.
Added new struct PLK_SecResource - needed by Somasung devices. It's not specialized struct (just general one), because I can't find any "sec" specification.

7pepo7 committed:
PATCH for Platinum library needed to external subtitles over UPnP works.

7pepo7 commented:
OK. If you could review it, Montellese.

I didn't change logic for searching best subtitles, cause in DVDPlayer it's based on streams, and here it's worked on files (on paths). I did'nt want to make changes in DVDPlayer class.

It's all should be changed when ProfileManager will be available for Kodi. Now subtitles are added in any available way, to support most of devices. But it should add subtitles only in one way (specific for target device), descibed in Profile for target device. I don't have too much free time, but I'll try to implement ProfileManager feature (but it could take some time). For now I think, that this solution will make some people happy (mostly no English-speaking).

Montellese commented:
Thanks for the adjustments. There are only a few minors left in the last commit.

7pepo7 commented:
Ok, I made adjustments. Thanks for guide me in git.

7pepo7 commented:
I think that now it's ok. Sorry for that. It's Visual Studio and its automatic tabulation.

Montellese commented:
Looks good for the next merge window. Thanks for taking the time to clean this up.

ace20022 commented:
Just a general note, preferred is misspelt.

7pepo7 committed:
It allows to play video with external subtitle over UPnP.

Supported subtitle file extensions: txt, srt, ssa, ass, sub, smi. (most popular ones)

Subtitle files are searched like in internal player (in video file folder and in folders specified in options). If more than one subtitle file are found (multiple language versions), Kodi takes the one with language choosed in options (prefered language for subtitles), otherwise first found.

Subtitles are added as 2 resources, 2 sec resources and 3 attributes added to video file resource, for greater compatibility with more UPnP devices. Device take the last one it could handle, and skip ones it can't "understand":

<res protocolInfo="http-get::video/mp4:" xmlns:pv="http://www.pv.com/pvns/" pv:subtitleFileUri="http://192.168.1.13:1722/video.srt" pv:subtitleFileType="SRT">http://192.168.1.13:1722/video.mp4 </res>
<res protocolInfo="http-get::text/srt:"> http://192.168.1.13:1722/video.srt </res>
<res protocolInfo="http-get::smi/caption:"> http://192.168.1.13:1722/video.srt </res>
<sec:CaptionInfoEx sec:type="srt">http://192.168.1.13:1722/video.srt</sec:CaptionInfoEx>
<sec:CaptionInfo sec:type="srt">http://192.168.1.13:1722/video.srt</sec:CaptionInfo>

Samsung devices get subtitles uri from "CaptionInfo.sec" field in http response header for movie file request (it contains "getCaptionInfo.sec" field).

This all should be (and probably it would be) changed, when Profile Manager (for DLNA devices) will be available in Kodi.

7pepo7 commented:
Thanks for notice, ace20022 (english it's not my primary language). Corrected.

Montellese commented:
jenkins build this please

ace20022 commented:
> Thanks for notice, ace20022

Thanks for this great contribution! As Montellese already said: "it'll make a lot of people happy" :+1:

MartijnKaijser commented:
@Montellese can we get your alpha2 milestone list merged as well?

Montellese commented:
@MartijnKaijser: I don't follow. All of the PRs I've assigned to the next merge window should be ready to be merged (and have been built by jenkins).

MartijnKaijser commented:
it's not worth releasing a first alpha1 while we still have frequent crashing on Android-x86 and Windows. So might as well already get the rest in and close merge window nearing ending of February. That way you can also go through your PRs more rapid.

Forage commented:
Not sure to what extend the bug tracker is being monitored for this feature, but this PR mean that at least http://trac.kodi.tv/ticket/15142 can be closed/updated.

NedScott commented:
Just a heads up, it seems this is causing quite a heavy load for some people's networks: http://forum.kodi.tv/showthread.php?tid=220840

Montellese commented:
@NedScott: Thanks for the hint. The logic should probably be adjusted to only look for external subtitles and provide them in the DIDL response if an item is starting to play. In all other cases the information about additional subtitles is not really of any use.

@7pepo7: Do you have time to look into this?

7pepo7 commented:
Ok. I'll take a look into this at saturday (sorry, no time at the moment).

Montellese commented:
I've taken a stab at this in https://github.com/Montellese/xbmc/commits/upnp_fix_ext_subtitles but I won't be able to test it myself but a forum user has volunteered. Will report back once I get some test results.
